### PR TITLE
[all] (Programming Guidance) Common coding issues and unexpected platform behaviors

### PR DESCRIPTION
**Pull request by AlexJerabek**
This documents some confusing edge-case situations our developers may encounter. I'm open to feedback on the structure and phrasing of the article. I'll also welcome additional scenarios to document.

Note that I'm still waiting on subject-matter expertise on the event handler issue. I can't figure out how to remove events handlers under any circumstance. I'll add the proper advice for developers once I learn it.

### CODE REVIEW

**Comment by o365devx** [comment]
Docs Build status updates of commit _[ccba092](https://github.com/OfficeDev/office-js-docs-pr/commits/ccba092b7f8ff946afc8a5b273a73e0246b45503)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/develop/common-coding-issues.md](https://github.com/OfficeDev/office-js-docs-pr/blob/AlexJ-IssuesPage/docs/develop/common-coding-issues.md) | :white_check_mark:Succeeded | |
[docs/toc.yml](https://github.com/OfficeDev/office-js-docs-pr/blob/AlexJ-IssuesPage/docs/toc.yml) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2019%5C10%5C24%5C57856ef7-e8cf-54ca-0d12-df5f42d1e8bb%5CPullRequest%5C201910241854563707-1387%5Cworkflow_report.html?sv=2016-05-31&sr=b&sig=gBiHZj4JtncZyXacmQhyeNEAT8pm2bc6qTfjYR%2BiuDI%3D&st=2019-10-24T18%3A51%3A15Z&se=2019-11-24T18%3A56%3A15Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Comment by AlexJerabek** [comment]
Tagging @sumurthy to take a look too.

File: docs/develop/common-coding-issues.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+```
+
+You can identify a property that must have its subproperties set with a JSON struct by checking its read-only modifier. All read-only properties can have their non-read-only subproperties directly set. Properties like `PageLayout.zoom` are not read-only and must be set with a JSON struct. In summary:
+
+- Read-only property: Subproperties can be set through navigation.
+- Writable property: Subproperties cannot be set through navigation.
+
+## Setting read-only properties
+
+The [TypeScript definitions](https://github.com/DefinitelyTyped/DefinitelyTyped/) for Office-JS specify which object properties are read-only. However, JavaScript developers are able to write code ignoring the `readonly` modifier. In that case, Office-JS silently ignores the write operation. The following example shows the read-only property [Chart.id](/javascript/api/excel/excel.chart#id) erroneously attempting to be set.
```
Comment (kbrandl) [inline docs/develop/common-coding-issues.md]: Nit: suggest rewording the final sentence ->

The following example erroneously attempts to set the read-only property [Chart.id](/javascript/api/excel/excel.chart#id).

File: docs/develop/common-coding-issues.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+// This will do nothing, since `id` is a read-only property.
+myChart.id = "5";
+```
+
+## Tracking event handlers
+
+> [!NOTE]
+> This section only applies to the host-specific APIs for Excel, OneNote, and Word.
+
+Event handlers are tied to individual, client-side, proxy objects. The event handler will attach to the related workbook object when synced. Removing an event handler requires a reference to the original proxy object.
```
Comment (kbrandl) [inline docs/develop/common-coding-issues.md]: Since the note says this section applies to Excel, OneNote, and Word -- probably should replace the word "workbook" with "document" instead.

File: docs/develop/common-coding-issues.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        var sameChart = context.workbook.worksheets.getActiveWorksheet().charts.getItem("Pie");
+        sameChart.onActivated.remove(chartActivated);
+        return context.sync();
+    });
+});
+```
+
+## See also
+
+- [The office-js GitHub repo](https://github.com/OfficeDev/office-js/issues): The issues page is a complete list of known product issues.
```
Comment (kbrandl) [inline docs/develop/common-coding-issues.md]:
```suggestion
- [OfficeDev/office-js](https://github.com/OfficeDev/office-js/issues): The place to report and view issues with the Office Add-ins platform and JavaScript APIs.
```

File: docs/develop/common-coding-issues.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+- [Stack Overflow](https://stackoverflow.com/questions/tagged/office-js): A question and answer site for professional and enthusiast programmers. Use the "office-js" tag when asking a Stack Overflow question so the community can find it and help.
```
Comment (kbrandl) [inline docs/develop/common-coding-issues.md]:
```suggestion
- [Stack Overflow](https://stackoverflow.com/questions/tagged/office-js): The place to ask and view programming questions about the Office JavaScript APIs. Be sure to apply the "office-js" tag to your question when posting to Stack Overflow.
```

File: docs/develop/common-coding-issues.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+- [User Voice](https://officespdev.uservoice.com/): A place to suggest new features for the Office platform.
```
Comment (kbrandl) [inline docs/develop/common-coding-issues.md]:
```suggestion
- [User Voice](https://officespdev.uservoice.com/): The place to suggest new features for the Office Add-ins platform and Office JavaScript APIs.
```

File: docs/develop/common-coding-issues.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+description: ''
+ms.date: 10/24/2019
+localization_priority: Normal
+---
+
+# Common coding issues and unexpected platform behaviors
+
+This articles highlights subtle discrepancies and unexpected behaviors with the Office JavaScript API. If you encounter an issue that belongs in this list, please let us know by using the GitHub feedback form at the bottom of the article.
+
+## Some properties cannot be set as navigational properties
```
[the remark this answers is not in the document]
Comment (AlexJerabek) [reply docs/develop/common-coding-issues.md]: There are still places we use that term (like the [ref docs](https://docs.microsoft.com/en-us/javascript/api/excel/excel.chart?view=excel-js-preview#load-propertynamesandpaths-). I'll link to the page defining that term. 

I agree that long-term we should remove the term "navigational".

**Comment by AlexJerabek** [comment]
Thanks @kbrandl. I made some changes based on your feedback.

**Comment by o365devx** [comment]
Docs Build status updates of commit _[26b3340](https://github.com/OfficeDev/office-js-docs-pr/commits/26b3340ab0cddd21f759f96349c2935350f404f1)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/develop/common-coding-issues.md](https://github.com/OfficeDev/office-js-docs-pr/blob/AlexJ-IssuesPage/docs/develop/common-coding-issues.md) | :white_check_mark:Succeeded | |
[docs/toc.yml](https://github.com/OfficeDev/office-js-docs-pr/blob/AlexJ-IssuesPage/docs/toc.yml) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2019%5C10%5C25%5C57856ef7-e8cf-54ca-0d12-df5f42d1e8bb%5CPullRequest%5C201910251944353787-1387%5Cworkflow_report.html?sv=2016-05-31&sr=b&sig=rJ8lW6lBTZNfACwsE%2FbcGgGYRUeVAJ%2Bbkuj%2FRrYi7UQ%3D&st=2019-10-25T19%3A41%3A01Z&se=2019-11-25T19%3A46%3A01Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

File: docs/develop/common-coding-issues.md [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        sameChart.onActivated.remove(chartActivated);
+        return context.sync();
+    });
+});
+```
+
+## See also
+
+- [[OfficeDev/office-js](https://github.com/OfficeDev/Office JS/issues): The place to report and view issues with the Office Add-ins platform and JavaScript APIs.
+- [Stack Overflow](https://stackoverflow.com/questions/tagged/Office JS): The place to ask and view programming questions about the Office JavaScript APIs. Be sure to apply the "office-js" tag to your question when posting to Stack Overflow.
```
Comment (kbrandl) [inline docs/develop/common-coding-issues.md]:
```suggestion
- [Stack Overflow](https://stackoverflow.com/questions/tagged/office-js): The place to ask and view programming questions about the Office JavaScript APIs. Be sure to apply the `office-js` tag to your question when posting to Stack Overflow.
```

File: docs/develop/common-coding-issues.md [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        var sameChart = context.workbook.worksheets.getActiveWorksheet().charts.getItem("Pie");
+        sameChart.onActivated.remove(chartActivated);
+        return context.sync();
+    });
+});
+```
+
+## See also
+
+- [[OfficeDev/office-js](https://github.com/OfficeDev/Office JS/issues): The place to report and view issues with the Office Add-ins platform and JavaScript APIs.
```
Comment (kbrandl) [inline docs/develop/common-coding-issues.md]:
```suggestion
- [OfficeDev/office-js](https://github.com/OfficeDev/Office JS/issues): The place to report and view issues with the Office Add-ins platform and JavaScript APIs.
```

**Comment by o365devx** [comment]
Docs Build status updates of commit _[25de328](https://github.com/OfficeDev/office-js-docs-pr/commits/25de328c5fe4697d41ad8c40bd778b38e819ebfb)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/develop/common-coding-issues.md](https://github.com/OfficeDev/office-js-docs-pr/blob/AlexJ-IssuesPage/docs/develop/common-coding-issues.md) | :white_check_mark:Succeeded | |
[docs/toc.yml](https://github.com/OfficeDev/office-js-docs-pr/blob/AlexJ-IssuesPage/docs/toc.yml) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2019%5C10%5C25%5C57856ef7-e8cf-54ca-0d12-df5f42d1e8bb%5CPullRequest%5C201910252307429430-1387%5Cworkflow_report.html?sv=2016-05-31&sr=b&sig=8Ay22a6XVEp0bqxEi2x3dn3r%2BALEebCLzCQL19g40og%3D&st=2019-10-25T23%3A04%3A04Z&se=2019-11-25T23%3A09%3A04Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Comment by o365devx** [comment]
Docs Build status updates of commit _[9f8c078](https://github.com/OfficeDev/office-js-docs-pr/commits/9f8c0782f017ef0b00408bf506331b584ba92645)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/develop/common-coding-issues.md](https://github.com/OfficeDev/office-js-docs-pr/blob/AlexJ-IssuesPage/docs/develop/common-coding-issues.md) | :white_check_mark:Succeeded | |
[docs/toc.yml](https://github.com/OfficeDev/office-js-docs-pr/blob/AlexJ-IssuesPage/docs/toc.yml) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2019%5C10%5C29%5C57856ef7-e8cf-54ca-0d12-df5f42d1e8bb%5CPullRequest%5C201910292221434895-1387%5Cworkflow_report.html?sv=2016-05-31&sr=b&sig=jT7dCoR8AMBMUq0HsSgRShuMmWR81Fgb%2FrXlr49lP44%3D&st=2019-10-29T22%3A17%3A58Z&se=2019-11-29T22%3A22%3A58Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Comment by AlexJerabek** [comment]
@kbrandl I've removed the event handler guidance from this article until I hear back from the engineering team. Are you okay with me checking the article in as is?

**Comment by kbrandl** [comment]
@AlexJerabek I'm fine with publishing this without the event handler guidance, but there are a couple of minor typos that should be fixed first in the **See also** section.

- remove extra bracket at the start of the first line
- "UserVoice" should be all one word (wierd, I know)

image

**Comment by o365devx** [comment]
Docs Build status updates of commit _[584447b](https://github.com/OfficeDev/office-js-docs-pr/commits/584447b1283cf245d7af7f87b551de1923448eb7)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/develop/common-coding-issues.md](https://github.com/OfficeDev/office-js-docs-pr/blob/AlexJ-IssuesPage/docs/develop/common-coding-issues.md) | :white_check_mark:Succeeded | |
[docs/toc.yml](https://github.com/OfficeDev/office-js-docs-pr/blob/AlexJ-IssuesPage/docs/toc.yml) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2019%5C10%5C29%5C57856ef7-e8cf-54ca-0d12-df5f42d1e8bb%5CPullRequest%5C201910292235167069-1387%5Cworkflow_report.html?sv=2016-05-31&sr=b&sig=ErVKb4U9bFTk2gzzNS3FufzWP2biPXQRBawmHRVUXAs%3D&st=2019-10-29T22%3A31%3A34Z&se=2019-11-29T22%3A36%3A34Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.
